Making unarchive idempotent

dagwieers opened the pull request:
##### ISSUE TYPE

<!-- Pick one below and delete the rest: -->
- Bugfix Pull Request
##### PLUGIN NAME

<!-- Name of the plugin/module/task  -->

unarchive
##### SUMMARY

<!-- Describe the change and the reason for it. -->

<!-- If you're fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does. -->

<!-- (Paste example execution or output here if necessary) -->

Originally unarchive was not idempotent and has many rough edges and bugs.
The current pull-request is a workable improvement on many fronts:
- zip support is now idempotent and fully functional
  - compares filetype, time, size, permission, ownership and CRC32
  - but gtar lacks check-mode, so check-mode is currently disabled
  - supports diff-mode (using rsync's itemized output format)
  - only extracts files that are in fact different (not all-or-nothing)
- gtar now better supports owner/group/mode changes (e.g. when provided as options, or as non-root)
- New option `exclude` to exclude specific paths/files
- New option `keep_newer` to exclude newer files on target
- New option `extra_opts` to influence unzip/gtar (like synchronize module)

This fixes #74, #348, #1292, #2037, #2250, #2438, #2491, #2772, #3229 and #3570 , but also fixes ansible/ansible#7188, ansible/ansible#7189 and ansible/ansible#9553.
(Please test your use-case again using this module)

The following items are part of future development (v2.1 ?):
- Re-implement the zip support using native zipfile module
- Re-implement the gtar support using native tarfile/gzip/bz2 modules (lzma/rarfile/cpio external)
- Implement diff-mode for gtar (using rsync's itemized output)
- Enable check-mode when gtar supports it (requires native re-implementation)
#### REVIEW AND TESTING

We need (more) feedback on the changes to the unarchive module. Please help review the code or test it in your environment. Testing is easy, simply perform the following in your playbook directory:

``` bash
mkdir -p ./library/
wget -O ./library/unarchive.py https://raw.githubusercontent.com/dagwieers/ansible-modules-core/idempotent-unarchive/files/unarchive.py
```
#### FUTURE NATIVE RE-IMPLEMENTATION APPROACH

The re-implementation of unzip/gtar support using native python modules will not only simplify the codebase, additional functionality can be implemented correctly and identically for all supported formats, which is currently not possible.

Whereas with the current implementation we cannot add specific functionality across archive formats, except by using non-generic `extra_opts` directives.

Native implementation will work for other platforms that may not have proper support/options that we rely one (e.g. Solaris, AIX, HP-UX).

Making unarchive idempotent is also nearly impossible using external commands. The current PR makes it possible for unzip, but gtar is a whole other beast to have a working check-mode. Same problem with unrar and other formats. Whereas if we use native libraries, we can abstract all layers natively and have a single implementation for all formats. In an idempotent fashion !
#### PYTHON 2.4

Since python 2.4 support is still needed for the next 2 years or so, we will have to implement some zipfile functionality ourselves on python 2.4, and possibly add lzma support directly into the module.

dagwieers committed:
WIP: Making unarchive idempotent

Currently unarchive is not idempotent and has many rough edges and bugs.
The current release is a workable improvement on many fronts:

- zip support is now idempotent (but gtar lacks check-mode)
- New option `exclude` to exclude specific paths/files
- New option `keep_newer` to exclude newer files on target
- New option `extra_opts` to influence unzip/gtar (like synchronize module)

The following items are still ongoing:

- Implement CRC32 support for .zip files
- Re-implement the zip support using native zipfile module
- Re-implement the gtar support using native tarfile/gzip/bz2 modules (lzma external)
- Implement check-mode (works in gzip, but fails using gtar)
- Implement diff-mode (discuss an appropriate output model, like synchronize module)

The re-implementation of unzip/gtar support using native python modules will not only simplify the codebase, additional functionality can be implemented correctly and identically, which is currently not possible. (Other archives could be implemented using native modules equally, incl. options)

pileofrogs commented:
For what little it may be worth, as the original author of unarchive, I give you my blessing to take over it's maintenance.  I really don't have time to re-implement the functionality of tar or zip, but if you do, my hat is off to you.

dagwieers committed:
Assorted fixes to zip support (during quality checks)

- Support both rw---- and rwx--- permstr
- Better file type support (more qa needed)
- Symlink support
- Include fix from #3229

dagwieers committed:
Implement zip diff-mode (itemized change) and avoid changes permissions every time (!)

This commit implements:
- rsync-compatible itemized-change output in diff-mode (using zip)
- no longer changing permissions unconditionally (when idempotent)

dagwieers committed:
Small fixes to itemized change output

dagwieers committed:
Fixes to user/group ownership changes

- The implementation of user/group ownership is a bit more complex for idempotency
- We report when a ZIP file incorrectly tags a directory as a file/link
- We only offer diff output when there is a change

dagwieers committed:
Fix the handling of includes and excludes for unzip

dagwieers committed:
Remove test output from output (confuses easily)

dagwieers committed:
Logic and performance improvements to ownership handling, and umask fix

dagwieers committed:
Handle special files (type '?')

gregdek commented:
Thanks @dagwieers for this PR. Unfortunately, it is not mergeable in its current state due to merge conflicts. Please rebase your PR. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

gregdek commented:
Thanks @dagwieers for this PR. This PR requires revisions, either because it fails to build or by reviewer request. Please make the suggested revisions. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

dagwieers committed:
Make exceptions compatible with python 2.4

dagwieers committed:
Implement CRC32 support

dagwieers commented:
@abadger @pileofrogs @gregdek Now that CRC32 support is implemented, I would like to get this into ansible-modules-core. A native re-implementation will require more time and effort (and may never materialize with python 2.4 support). But this implementation is an improvement over the existing unarchive module, especially for zip-support and additional options.

Let me know what is required to get this upstreamed. It is priority 1 for me (and would fix a few reported issues/feature requests as well).

dagwieers committed:
Revert some unintended/unknown changes ?

dagwieers committed:
Taking over maintenance as offered by current maintainer

abadger commented on files/unarchive.py:
```diff
@@ -89,46 +107,358 @@
+OWNER_DIFF_RE = re.compile(r': Uid differs$')
+GROUP_DIFF_RE = re.compile(r': Gid differs$')
+MODE_DIFF_RE = re.compile(r': Mode differs$')
+MODE_DIFF_RE = re.compile(r' is newer or same age.$')
```
Different name needed for this regex?

dagwieers replied:
To be honest, these changes do not really make a difference (pun intended). I can remove those changes from this PR. This was an attempt to implement check-mode using gtar, but I think a native re-implementation is in order. Shall I remove those ?

dagwieers replied:
After more testing it seems those changes do fix a problem. gtar --diff compares owner/group information against the content of the archive, which will induce a (false) change when you run it as non-root (or when owner/group is supplied).

A rewrite can properly implement what is needed/expected, similar to what we now do for zip-support.

dagwieers replied:
I now see what you mean. Fixed that.

dagwieers committed:
Fix support for white-spaces in filenames

dagwieers committed:
Remove/rename incorrect regex

dagwieers committed:
Ensure that fat executables end up with execute permission

dagwieers committed:
Remove check_result from output when unchanged

dagwieers committed:
When unarchiving as a user, or when owner/group/mode is supplied --diff is insufficient

Only way to be sure is to check request with what is on disk (as we do for zip).
Leave this up to set_fs_attributes_if_different() instead of inducing a (false) change

dagwieers committed:
By default, don't send confusing check_results in verbose output

This fixes #74.

ghost commented:
I have exactly right scenario for this, I'll test this out.

dagwieers commented:
ready_for_review

gregdek commented:
Thanks @dagwieers. To the current maintainers, @pileofrogs please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with text 'shipit' or 'needs_revision' as appropriate.

[This message brought to you by your friendly Ansibull-bot.]

pileofrogs commented:
I'm sorry to say I seem to take longer responding to these than I should.
Am I holding things up?  If so, is there a way to get myself out of the
critical path?

On Thu, Mar 31, 2016 at 7:29 AM, Greg DeKoenigsberg <
notifications@github.com> wrote:

> Thanks @dagwieers https://github.com/dagwieers. To the current
> maintainers, @pileofrogs https://github.com/pileofrogs please review
> according to guidelines (
> http://docs.ansible.com/ansible/developing_modules.html#module-checklist)
> and comment with text 'shipit' or 'needs_revision' as appropriate.
> 
> [This message brought to you by your friendly Ansibull-bot.]
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> https://github.com/ansible/ansible-modules-core/pull/3307#issuecomment-203965062

gregdek commented:
Thanks again to @dagwieers for this PR, and thanks @dagwieers, @pileofrogs for reviewing. Marking for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

gregdek commented:
Thanks again to @dagwieers. This module is going into community review.

We notice that you are one of the maintainers (@dagwieers, @pileofrogs) of this module, so when you think it's ready, please comment "shipit" and we will consider it for merging.

[This message brought to you by your friendly Ansibull-bot.]

dagwieers commented:
shipit

gregdek commented:
Thanks @dagwieers. Since you are one of the maintainers (@dagwieers, @pileofrogs) of this module and you commented with "shipit", we are marking this PR for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

gregdek commented:
@pileofrogs we've added @dagwieers as a reviewer for this module, so now you've got some backup. :) I'd like to keep you as a maintainer, though, if that's ok.

gregdek commented:
Thanks again to @dagwieers. This module is going into community review.

We notice that you are one of the maintainers (@dagwieers, @pileofrogs) of this module, so when you think it's ready, please comment "shipit" and we will consider it for merging.

[This message brought to you by your friendly Ansibull-bot.]

dsKarthick commented:
Having `extra_opts` is extremely useful. Can't wait to have it.

gregdek commented:
Hm, looks like this got moved out of shipit for some reason. Putting back into shipit. We'll see if the bot misbehaves again.

gregdek commented:
Thanks again to @dagwieers. This module is going into community review.

We notice that you are one of the maintainers (@dagwieers, @pileofrogs) of this module, so when you think it's ready, please comment "shipit" and we will consider it for merging.

[This message brought to you by your friendly Ansibull-bot.]

gregdek commented:
OK, so Ansibullbot keeps screwing this one up. Why is unclear.

So here's what I'm going to do: I'm going to start whacking comments! We'll see if that fixes the problem.

gregdek commented:
It didn't, so I have instead implemented a bot change. I think it worked. We'll see!

Jmainguy commented:
@dagwieers Think you could squash the 19 commits into one? Would make it easier for people to see whats going on in the future when this gets merged.

dagwieers commented:
@Jmainguy What's the point of squashing the commits ?
You can see what's going on when you click on: [Files changed](https://github.com/ansible/ansible-modules-core/pull/3307/files) at the top, no ?

bcoca commented:
squashing makes it easy to handle the commit (cherry-pick/revert/bisect/etc), but really not needed since now github allows us to do this through the merge button

dagwieers commented:
Alright, if anyone cares about squash/merge: https://github.com/blog/2141-squash-your-commits
Explained well...

sunshine69 commented:
The patch gives error when I am trying to unzip the Pentaho archive. pdi-ce-6.0.0.0-353.zip - if anyone wants more debugging I will post the file in origin - it is pretty big though (780MB).

My task is simple like this

```

- name: Install pentaho
  unarchive: copy=no src=/tmp/{{PDI_ARCHIVE}} dest=/opt/pentaho
  become: true
  become_user: 'jumbo'
```

The exact same task runs fine with the existing unarchive module but it gives a lot of non-relevant error message about gtar (it is extracted properly).

The error message is:

"msg": "Unexpected error when accessing exploded file: [Errno 2] No such file or directory: '/opt/pentaho/data-integration/Data Integration.app/'"

dagwieers commented:
Please test the module in the master branch.

sunshine69 commented:
HI, does it mean the below 

```
mkdir -p ./library/
wget -O ./library/unarchive.py https://raw.githubusercontent.com/dagwieers/ansible-modules-core/idempotent-unarchive/files/unarchive.py
```

does not work?
The latest ansible does not work for me for some reason and we have no time to evaluate it again. Currently we use a fixed version 2.0.0.2 and the above is best way for me to test and use the backported fix.

If it requires latest version of ansible I can try again only for that task and see if problem is fixed.
